### PR TITLE
fix: parse jfr to pprof, without intermediate tree

### PR DESCRIPTION
**Pull request by korniltsev**
```
goos: darwin
goarch: arm64
pkg: github.com/grafana/pyroscope/pkg/ingester/pyroscope
                                                               │   old.txt    │               new.txt               │
                                                               │    sec/op    │    sec/op     vs base               │
IngestJFR/cortex-dev-01__kafka-0__cpu__0.jfr.gz-8                3.284m ± ∞ ¹   3.234m ± ∞ ¹   -1.54% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu__1.jfr.gz-8                3.892m ± ∞ ¹   3.952m ± ∞ ¹        ~ (p=1.000 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu__2.jfr.gz-8                3.298m ± ∞ ¹   3.999m ± ∞ ¹  +21.25% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu__3.jfr.gz-8                4.081m ± ∞ ¹   4.806m ± ∞ ¹  +17.78% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu_lock0_alloc0__0.jfr.gz-8   24.74m ± ∞ ¹   23.59m ± ∞ ¹   -4.66% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu_lock_alloc__0.jfr.gz-8     6.908m ± ∞ ¹   7.267m ± ∞ ¹   +5.20% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu_lock_alloc__1.jfr.gz-8     6.620m ± ∞ ¹   7.038m ± ∞ ¹   +6.31% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu_lock_alloc__2.jfr.gz-8     6.962m ± ∞ ¹   7.269m ± ∞ ¹   +4.40% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu_lock_alloc__3.jfr.gz-8     6.402m ± ∞ ¹   6.685m ± ∞ ¹   +4.41% (p=0.029 n=4)
geomean                                                          5.901m         6.243m         +5.79%
¹ need >= 6 samples for confidence interval at level 0.95

                                                               │    old.txt    │               new.txt                │
                                                               │     B/op      │     B/op       vs base               │
IngestJFR/cortex-dev-01__kafka-0__cpu__0.jfr.gz-8                2.865Mi ± ∞ ¹   2.170Mi ± ∞ ¹  -24.26% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu__1.jfr.gz-8                3.271Mi ± ∞ ¹   2.357Mi ± ∞ ¹  -27.95% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu__2.jfr.gz-8                2.848Mi ± ∞ ¹   2.126Mi ± ∞ ¹  -25.36% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu__3.jfr.gz-8                3.365Mi ± ∞ ¹   2.429Mi ± ∞ ¹  -27.82% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu_lock0_alloc0__0.jfr.gz-8   17.57Mi ± ∞ ¹   13.79Mi ± ∞ ¹  -21.55% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu_lock_alloc__0.jfr.gz-8     5.667Mi ± ∞ ¹   3.590Mi ± ∞ ¹  -36.64% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu_lock_alloc__1.jfr.gz-8     5.489Mi ± ∞ ¹   3.519Mi ± ∞ ¹  -35.89% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu_lock_alloc__2.jfr.gz-8     5.782Mi ± ∞ ¹   3.647Mi ± ∞ ¹  -36.93% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu_lock_alloc__3.jfr.gz-8     5.319Mi ± ∞ ¹   3.386Mi ± ∞ ¹  -36.34% (p=0.029 n=4)
geomean                                                          4.859Mi         3.375Mi        -30.54%
¹ need >= 6 samples for confidence interval at level 0.95

                                                               │   old.txt    │               new.txt               │
                                                               │  allocs/op   │  allocs/op    vs base               │
IngestJFR/cortex-dev-01__kafka-0__cpu__0.jfr.gz-8                39.30k ± ∞ ¹   32.55k ± ∞ ¹  -17.18% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu__1.jfr.gz-8                43.38k ± ∞ ¹   35.66k ± ∞ ¹  -17.81% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu__2.jfr.gz-8                38.88k ± ∞ ¹   31.74k ± ∞ ¹  -18.36% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu__3.jfr.gz-8                44.33k ± ∞ ¹   36.57k ± ∞ ¹  -17.51% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu_lock0_alloc0__0.jfr.gz-8   139.2k ± ∞ ¹   138.3k ± ∞ ¹        ~ (p=0.343 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu_lock_alloc__0.jfr.gz-8     66.38k ± ∞ ¹   49.91k ± ∞ ¹  -24.80% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu_lock_alloc__1.jfr.gz-8     64.72k ± ∞ ¹   49.46k ± ∞ ¹  -23.58% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu_lock_alloc__2.jfr.gz-8     68.10k ± ∞ ¹   51.86k ± ∞ ¹  -23.86% (p=0.029 n=4)
IngestJFR/cortex-dev-01__kafka-0__cpu_lock_alloc__3.jfr.gz-8     62.49k ± ∞ ¹   47.30k ± ∞ ¹  -24.32% (p=0.029 n=4)
geomean                                                          58.05k         47.04k        -18.96%
```
- [x] still do pprof->bytes roundtrip in this pr, improve in a followup 
- [x] add testdata for profile id
- [x] add testdata for wall
- [x] add testdata with labels for ingest handler

Future improvements:
- improve `mergeJVMGeneratedClasses`
- implement more close to e2e test - ingest to Head, query back

### CODE REVIEW

File: pkg/distributor/distributor.go [hunk 1]
```diff
@@ -177,26 +179,76 @@ func (d *Distributor) stopping(_ error) error {
 	return services.StopManagerAndAwaitStopped(context.Background(), d.subservices)
 }
 
-func (d *Distributor) Push(ctx context.Context, req *connect.Request[pushv1.PushRequest]) (*connect.Response[pushv1.PushResponse], error) {
+func (d *Distributor) Push(ctx context.Context, grpcReq *connect.Request[pushv1.PushRequest]) (*connect.Response[pushv1.PushResponse], error) {
```
Comment (cyriltovena) [inline pkg/distributor/distributor.go:182]: Why did you need to change this ?

Comment (korniltsev) [reply pkg/distributor/distributor.go:182]: I want to be able to Push a parsed pprof and I dont see how I could do it with grpc.

I want to be able Push `[]byte` with grpc and `*pprof.Profile` with a direct function call
So there's new `ProfileSample` which has both and sometimes `RawProfile []byte` is nil (for a case when pprof was not parsed from bytes, i.e. jfr)

I also want to have a single function processing both cases avoiding code duplication.

Do you think there may be other ways to implement this? I'd like to improve it

File: pkg/distributor/distributor.go [hunk 2]
```diff
@@ -207,19 +259,23 @@ func (d *Distributor) Push(ctx context.Context, req *connect.Request[pushv1.Push
 		for _, raw := range series.Samples {
 			usagestats.NewCounter(fmt.Sprintf("distributor_profile_type_%s_received", profName)).Inc(1)
 			d.profileReceivedStats.Inc(1)
-			d.bytesReceivedTotalStats.Inc(int64(len(raw.RawProfile)))
-			d.bytesReceivedStats.Record(float64(len(raw.RawProfile)))
+			if haveRawPprof {
+				d.metrics.receivedCompressedBytes.WithLabelValues(profName, tenantID).Observe(float64(len(raw.RawProfile)))
```
Comment (cyriltovena) [inline pkg/distributor/distributor.go:263]: it makes this metrics a bit useless. 

We should try to improve it, may be this should not be supposed to be recorded here ?

File: pkg/distributor/distributor.go (hunk 1 above)
Comment (cyriltovena) [reply pkg/distributor/distributor.go:182]: Not sure but I'll provide feedback, I think it makes senses.

File: pkg/distributor/distributor.go [hunk 3]
```diff
@@ -207,19 +259,23 @@ func (d *Distributor) Push(ctx context.Context, req *connect.Request[pushv1.Push
...
-			d.metrics.receivedCompressedBytes.WithLabelValues(profName, tenantID).Observe(float64(len(raw.RawProfile)))
-			p, err := pprof.RawFromBytes(raw.RawProfile)
-			if err != nil {
-				return nil, connect.NewError(connect.CodeInvalidArgument, err)
+			p := raw.Profile
+			var decompressedSize int
+			if haveRawPprof {
+				decompressedSize = p.SizeBytes()
+			} else {
+				decompressedSize = p.SizeVT()
```
Comment (cyriltovena) [inline pkg/distributor/distributor.go:271]: What's the difference between `p.SizeBytes()` and `p.SizeVT()`. Sounds like SizeVT is more accurate ? If so may be we should always use it.

Comment (korniltsev) [reply pkg/distributor/distributor.go:271]: the difference is `SizeBytes() ` is just `len(buf)` and `SizeVT` does the deep structural size computation
for jfr `len(buf) == 0`

File: pkg/distributor/distributor.go (hunk 1 above)
Comment (cyriltovena) [reply pkg/distributor/distributor.go:182]: One thing that could help either having 2 commits, or 2 prs too for both change, the jfr improvement and the parsed pprof change.

Comment (korniltsev) [reply pkg/distributor/distributor.go:182]: Yes, let's do two PR's!

Comment (korniltsev) [reply pkg/distributor/distributor.go:182]: But I need an idea of how to implement the second one before I start

File: pkg/distributor/distributor.go (hunk 2 above)
Comment (korniltsev) [reply pkg/distributor/distributor.go:263]: I dont get it. Why is it useless?
It is just a little bit less accurate for JFR because there is no such thing as `compressedBytes(profName="memory")` because memory is compressed together with cpu/lock/... so we count this metric for JFR as a single profName="JFR"

File: pkg/distributor/distributor.go (hunk 3 above)
Comment (korniltsev) [reply pkg/distributor/distributor.go:271]: so we need to calculate it

File: pkg/distributor/distributor.go (hunk 1 above)
Comment (korniltsev) [reply pkg/distributor/distributor.go]: reverted changes in distributor

File: pkg/distributor/distributor.go (hunk 2 above)
Comment (cyriltovena) [reply pkg/distributor/distributor.go]: I meant that the metrics now doesn't account for all ingested profile. so I wonder if we should deprecated at some point or change the way it works.

File: pkg/distributor/distributor.go (hunk 3 above)
Comment (cyriltovena) [reply pkg/distributor/distributor.go]: yeah I was wondering how much overhead SizeVT() and if we should consider just calling it all the time. I think it's fine to keep how you did.

File: pkg/distributor/distributor.go (hunk 2 above)
Comment (korniltsev) [reply pkg/distributor/distributor.go]: it does account for all ingested profiles, just less accurate for jfr(no split by profile type)

File: pkg/distributor/distributor.go (hunk 3 above)
Comment (korniltsev) [reply pkg/distributor/distributor.go]: resolving this as I reverted distributor changes

File: pkg/ingester/pyroscope/ingest_adapter.go [hunk 4]
```diff
@@ -152,6 +159,55 @@ func (p *pyroscopeIngesterAdapter) Evaluate(input *storage.PutInput) (storage.Sa
...
+	if err != nil {
+		return fmt.Errorf("parsing IngestInput-pprof failed %w", err)
+	}
+	grpcReq, err := p.convertToGRPC(plainReq)
+	if err != nil {
+		return fmt.Errorf("converting IngestInput-pprof failed %w", err)
+	}
+	_, err = p.svc.Push(ctx, grpcReq)
+	if err != nil {
+		return fmt.Errorf("pushing IngestInput-pprof failed")
```
Comment (cyriltovena) [inline pkg/ingester/pyroscope/ingest_adapter.go:173]: looks like you're swallowing the error here ? Which mean this will always translate into 500 and should not because  there's more inside this error like validation and throttling.

File: pkg/ingester/pyroscope/ingest_adapter.go [hunk 5]
```diff
@@ -152,6 +159,55 @@ func (p *pyroscopeIngesterAdapter) Evaluate(input *storage.PutInput) (storage.Sa
 	return nil, false // noop
 }
 
+func (p *pyroscopeIngesterAdapter) parseToPprof(ctx context.Context, in *ingestion.IngestInput, pprofable ingestion.ParseableToPprof) error {
+	plainReq, err := pprofable.ParseToPprof(ctx, in.Metadata)
```
Comment (cyriltovena) [inline pkg/ingester/pyroscope/ingest_adapter.go:163]: Anything in that err should translate into 4xx eg. if JFR can't be parsed...

**Review by cyriltovena — APPROVED** [review]
LGTM

File: pkg/ingester/pyroscope/ingest_adapter.go (hunk 5 above)
Comment (korniltsev) [reply pkg/ingester/pyroscope/ingest_adapter.go:163]: This is exactly like that. I will add a unit test to double check

File: pkg/ingester/pyroscope/ingest_adapter.go (hunk 4 above)
Comment (korniltsev) [reply pkg/ingester/pyroscope/ingest_adapter.go]: fixed

File: pkg/ingester/pyroscope/ingest_adapter.go (hunk 5 above)
Comment (korniltsev) [reply pkg/ingester/pyroscope/ingest_adapter.go:163]: https://github.com/grafana/pyroscope/pull/2355/commits/5970622861a163d6cec54677feec6fe3b7482e64
